zebra-chain/src/orchard/tree.rs

```diff
@@ -292,6 +295,71 @@ pub struct NoteCommitmentTree {
     cached_root: std::sync::RwLock<Option<Root>>,
 }
 
+/// The block height of the final leaf in a level 16 subtree and
+/// the block height of the first leaf of the next subtree
+#[derive(Clone, Debug)]
+pub struct NoteCommitmentSubtree {
+    /// Index of this subtree
+    pub index: u16,
+    /// Start boundary of this subtree, the block height of its last leaf.
+    pub start: Option<block::Height>,
+    /// End boundary of this subtree, the block height of its last leaf.
+    pub end: block::Height,
+    /// Whether this subtree's end height is the start height of the next subtree.
+    pub begins_next_subtree: Option<bool>,
```

Comment:
There should only be some `start` height when reading from disk, and it should only have `begins_new_subtree` when updating note commitment trees. When a block finishes a subtree without starting the next one, the start height for the next subtree is the height of the next note.

Reply:
> There should only be some start height when reading from disk, and it should only have begins_new_subtree when updating note commitment trees.

After reading the docs and your explanation, I'm not quite sure what these fields do or how they are used. As far as I can see, each note commitment subtree should always have a start height. So if there's no start height, the entire subtree could be `None`?

Nit: These details seem like something useful to document.

Optional: It seems like you're trying to implement multiple different pieces of functionality using the same struct. Would an enum (or a struct containing an enum, or an enum containing structs) be a better way to represent those states explicitly?

For example, you might end up with something like:
```rust
enum NoteCommitmentSubtreeState {
    Pending { index }
    Started { index, start }
    Completed { index, start, end, node }
    CompletedOnDisk { index, start, end, begins_next_subtree, node }
}
```

Reply:
Is the enum necessary if we are only storing the index, end height, and node?

Is there a way to avoid this state machine (and the associated code complexity) by getting the completed subtree from the existing frontier in the note commitment tree?